docs: refactor local ZK stack docs (#176)
matter-labs/zksync-docs: content/10.zk-stack/20.running-a-zk-chain/30.custom-base-tokens.md

---
title: Custom Base Tokens
description:
---

You can customize your ZK chain by changing the base token used for gas from ETH to an ERC20 token.

> With the initial release of ZK stack, base tokens must be whitelisted for use as a base token for a chain.

Reviewer: ```suggestion
To use a token as a base token on a ZK chain, it must be whitelisted in the `BridgeHub` contract on the L1.
```
A base token is only a base token if it is whitelisted, even if it is ETH? Is ETH the default base token and it doesn't need to be whitelisted, or is this a process for any ERC20 token that is going to be a base token?

Author: it's just for ERC20 tokens afaik

> The whitelisted addresses are stored and checked in the `BridgeHub` contract on the L1.

Reviewer: _I'm still new to ZK chains so my knowledge is almost non-existent, so my questions might be dumb questions._

By "addresses" this means that there can be more than one base token for a chain? Or that every chain's base token address is available on the L1?

Author: I might also be wrong here, but I think every chain's base token address is available on the L1 through this `BridgeHub` contract

In the future it will be possible to add a new token in a permissionless process.

> For now, you have the ability to whitelist tokens in your local ecosystem.

Reviewer 2: suggest allowlist instead

## Custom Base Token Setup

> The overall flow for setting up a local chain with a custom ERC20 base token looks like this:

Reviewer: ```suggestion
The overall flow for setting up a local chain with a custom ERC20 base token looks like the following:
```

1. Send ERC20 tokens to both the ecosystem and chain governor addresses on the L1.
    - You can find the addresses for the governor in `path/to/elastic_chain_ecosystem/configs/wallets.yaml`
    and `path/to/elastic_chain_ecosystem/chains/my_chain/configs/wallets.yaml` under `governor`.
    - See the example below for how to do this using the `cast` CLI:

    ```bash
    cast send <0xYOUR_TOKEN_ADDRESS> "transfer(address,uint256)" \
    <0x_YOUR_GOVERNOR_ADDRESS> 1000000000000000000 \
    --private-key <0x_YOUR_PRIVATE_KEY> \
    --rpc-url http://localhost:8545 \
    --gas-price 30000000000
    ```

1. Create your new chain that uses the ERC20 token and set it as the default.
    - When you set up your chain, you will be asked for the nominator and denominator for your base token.
    - The base token nominator and denominator are used to define the relationship of the base token price with ETH.
    - For example, if you set the nominator to 20 and the denominator to 1, together the relation is 20/1.
    This would mean that 20 tokens would be given the equivalent value as 1 ETH for gas.
1. Initialize the new chain in the ecosystem.
1. Bridge tokens from the L1 to your new chain.

>     - You can use the `zksync-cli` to bridge the tokens:

Reviewer: ```suggestion
    - You can use ZKsync CLI to bridge the tokens:
```

    ```bash

>     npx zksync-cli bridge deposit --token <0x_YOUR_TOKEN_ADDRESS> \

Reviewer: ```suggestion
    zksync-cli bridge deposit --token <0x_YOUR_TOKEN_ADDRESS> \
```

    --rpc=http://localhost:3050 \
    --l1-rpc=http://localhost:8545
    ```

When you bridge the ERC20 token to your ZK chain, the token resolves to the same address used for ETH on a standard ZK chain:

```ts
const L2_BASE_TOKEN_ADDRESS = '0x000000000000000000000000000000000000800a';
```

> So if you want to check the balance of your bridged tokens with the `zksync-cli`, you don't need to input the token address.

Reviewer: ```suggestion
If you want to check the balance of your bridged tokens with ZKsync CLI, you don't need to input the token address.
```

```bash
zksync-cli wallet balance --address <0x_YOUR_ADDRESS> --rpc http://localhost:3050
```

## Restarting a custom base token chain

To run your custom ZK chain again after setting up and shutting it down, follow the steps below:

1. If you shut down your local L1 node that had your ERC20 token deployed to it:
    - Redeploy your ERC20 contract to the L1.
    - Update the base token address in `my_elastic_chain/chains/custom_zk_chain/configs/contracts.yaml` under `l1.base_token_addr` and in

>   `my_elastic_chain/chains/custom_zk_chain/ZkStack.yaml` under `base_token.address`.

Reviewer: Is the `my_elastic_chain` an actual name in setting up a ZK chain project, or is that the name of the ZK chain you created? 

For examples like `path/to/elastic_chain_ecosystem` I think we should put them in `< >` brackets to help clearly define them as variables for the user to change. https://developers.google.com/style/placeholders

Or we clearly write out "in your ZK chain project directory, update the base token address in `/chains/custom_zk_chain/configs/contracts.yaml`"

```suggestion
  `<zk chain project directory>/chains/custom_zk_chain/ZkStack.yaml` under `base_token.address`.
```

Author: good catch, these are just names I made up
